### ISSUE TITLE
Elastic Horovod - Tensorflow/Keras - New worker hangs on start up

### ISSUE THREAD

**Issue by czmrand** [horovod/horovod#2652]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 2.3.1
3. Horovod version: 0.21.1 (installed with HOROVOD_WITH_GLOO=1)
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version: 3
8. Spark / PySpark version:
9. OS and version:
10. GCC version:
11. CMake version:

I am attempting to run the attached script on 3 systems.
1. When I remove one of the workers (either by stopping the machine or killing the process), the training continues on the two remaining workers (as advertised) but I get the following error message:
[0]<stderr>:[2021-02-08 09:24:59.555830: E /tmp/pip-install-7tlcippf/horovod_015e4deaf97a4b33908ab119077a4f9c/horovod/common/operations.cc:548] Horovod background loop uncaught exception: [/tmp/pip-install-7tlcippf/horovod_015e4deaf97a4b33908ab119077a4f9c/third_party/compatible_gloo/gloo/transport/tcp/pair.cc:566] Read error [10.0.1.246]:63386: Connection reset by peer
Is this expected?

2. The addition of a worker is also identified successfully, however the new worker stalls just before starting the train loop. As a result, the training on the other workers stalls as well with:
[0]<stderr>:0: [HorovodBroadcast]
[0]<stderr>:1: [HorovodBroadcast]
[0]<stderr>:2: [HorovodBroadcast_Adam_conv1_bn_beta_m_0, HorovodBroadcast_Adam_conv1_bn_beta_v_0, HorovodBroadcast_Adam_conv1_bn_gamma_m_0, HorovodBroadcast_Adam_conv1_bn_gamma_v_0, HorovodBroadcast_Adam_conv1_conv_bias_m_0, HorovodBroadcast_Adam_conv1_conv_bias_v_0 ...]
[0]<stderr>:[2021-02-08 10:54:33.732016: W /tmp/pip-install-7tlcippf/horovod_015e4deaf97a4b33908ab119077a4f9c/horovod/common/stall_inspector.cc:105] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
[0]<stderr>:Missing ranks:

[code.tar.gz](https://github.com/horovod/horovod/files/5943484/code.tar.gz)

[data.tar.gz](https://github.com/horovod/horovod/files/5943469/data.tar.gz)

import argparse
import tensorflow as tf
import horovod.tensorflow.keras as hvd

hvd.init()

gpus = tf.config.experimental.list_physical_devices('GPU')
for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)
if gpus:
    tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], 'GPU')

def get_dataset(batch_size):
    def parse_image_function(example_proto):
        image_feature_description = {
            'image': tf.io.FixedLenFeature([], tf.string),
            'label': tf.io.FixedLenFeature([], tf.int64)
        }
        features = tf.io.parse_single_example(example_proto, image_feature_description)
        image = tf.io.decode_raw(features['image'], tf.uint8)
        image.set_shape([3 * 32 * 32])
        image = tf.reshape(image, [32, 32, 3])
        label = tf.cast(features['label'], tf.int32)
        return image, label

    from tensorflow.keras.layers.experimental import preprocessing

    def rescale(image, label):
        image = preprocessing.Rescaling(1.0 / 255)(image)
        return image, label

    autotune = tf.data.experimental.AUTOTUNE
    options = tf.data.Options()
    options.experimental_deterministic = False
    records = tf.data.Dataset.list_files('data/*', shuffle=True).with_options(options)
    ds = tf.data.TFRecordDataset(records, num_parallel_reads=autotune).repeat()
    ds = ds.map(parse_image_function, num_parallel_calls=autotune)
    ds = ds.batch(batch_size)
    ds = ds.map(rescale,num_parallel_calls=autotune)
    ds = ds.prefetch(autotune)
    return ds

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Train resnet50 cifar10")
    parser.add_argument("--lr", type=float, default=0.001)
    parser.add_argument("--model_dir", type=str, default="./model_keras_resnet")

    args = parser.parse_args()

    ds = get_dataset(batch_size=2)

    model = tf.keras.applications.resnet.ResNet50(weights=None, input_shape=(32, 32, 3), classes=10)
    model.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(),
                  optimizer=tf.keras.optimizers.Adam())

    def on_state_reset():
        print('-----------------------------------')
        print('resetting learning rate')
        tf.keras.backend.set_value(model.optimizer.lr, 0.001)
        print('-----------------------------------')

    state = hvd.elastic.KerasState(model, batch=100, epoch=0)
    state.register_reset_callbacks([on_state_reset])

    callbacks = [
        hvd.elastic.CommitStateCallback(state),
        hvd.elastic.UpdateBatchStateCallback(state),
        hvd.elastic.UpdateEpochStateCallback(state),
    ]

    @hvd.elastic.run
    def train(state):
        model.fit(ds, steps_per_epoch=100, epochs=100, callbacks=callbacks)
    train(state)

**Comment by tgaddair** [comment]
Thanks for providing a repro script @czmrand.  I will take a look.

**Comment by chongxiaoc** [comment]
Here is what I saw from reproducing the bug:
- Since you only have 3 hosts, 1 host is down, it is added into blacklist, for example, from my log stdout:
```
DEBUG:root:host worker-2 has been blacklisted, ignoring exit from local_rank=0
```
- After shrinking, training continues to run on two hosts, that's expected.
- Then you added the `blacklisted` host back, in this case, horovod should not count this host in, see:
https://github.com/horovod/horovod/blob/6916985c9df111f36864724e2611827f64de8e11/horovod/runner/elastic/discovery.py#L138

- However, in the following training, the blackedlist host is counted into collective as a missing rank.

I think something is wrong in discovery blacklisted hosts part.

**Comment by chongxiaoc** [comment]
Update, part of my above comment is incorrect, since I was inject failures by removing/adding hosts in hostfile, it is a graceful removal, the removed host should not be blacklisted.

The output `DEBUG:root:host worker-2 has been blacklisted, ignoring exit from local_rank=0` actually comes from no remaining slot for the removed host from https://github.com/horovod/horovod/blob/32a84f60a95560483e05e717e0e9576e354a97c8/horovod/runner/elastic/driver.py#L139

In this case, `worker-2` is legal to be added back, but I'm also seeing the hanging there.

**Comment by chongxiaoc** [comment]
@czmrand 
I finally fixed the issue.
Please see the attached training scripts, and the diff is as below:
```
(hvd-py3.7) chongxiaoc@chongxiaoc-C02DL7WPMD6R hvd_sandbox % diff train_gloo.py train_gloo_fixed.py
62c62,66
<     state = hvd.elastic.KerasState(model, batch=100, epoch=0)
---
>
>     # Horovod: initialize model and optimizer state so we can synchronize across workers
>     model.fit(ds, steps_per_epoch=1, epochs=1, callbacks=None)
>
>     state = hvd.elastic.KerasState(model, batch=0, epoch=0)
66,67d69
<         hvd.elastic.CommitStateCallback(state),
<         hvd.elastic.UpdateBatchStateCallback(state),
68a71,72
>         hvd.elastic.UpdateBatchStateCallback(state),
>         hvd.elastic.CommitStateCallback(state),
71d74
<
74c77
<         model.fit(ds, steps_per_epoch=100, epochs=100, callbacks=callbacks)
---
>         state.model.fit(ds, steps_per_epoch=100, epochs=100-state.epoch, callbacks=callbacks)
```

Some import points you were missing:
- Keras/TF2 has a lazy initialization for optimizer weights (https://sourcegraph.com/github.com/tensorflow/tensorflow@v2.4.1/-/blob/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L351:10),  so hovorod has to initialize optimizer state by running one step of training so it can synchronize across all ranks, you can refer it by checking `tensorflow2_mnist_elastic.py` https://github.com/horovod/horovod/blob/f3af98649a26f4f3725fcab6a9bd3e8d29d7ffd2/examples/elastic/tensorflow2/tensorflow2_mnist_elastic.py#L70

Your above example hangs there, because the new joined rank has an empty optimizer state to sync with existing rank 0 and rank1, i.e., 
```
[0]:2: [HorovodBroadcast_Adam_conv1_bn_beta_m_0, HorovodBroadcast_Adam_conv1_bn_beta_v_0, HorovodBroadcast_Adam_conv1_bn_gamma_m_0, HorovodBroadcast_Adam_conv1_bn_gamma_v_0, HorovodBroadcast_Adam_conv1_conv_bias_m_0, HorovodBroadcast_Adam_conv1_conv_bias_v_0 ...]
```
There are all optimizer variables.

- Also I adjusted a few places of using elastic state and callback functions order, very minor changes.

Now the training script is running fine on my system by injecting a failure of using removal hosts, and also adding hosts back into training works fine, no hanging anymore.

I would like to take this chance of fixing your example, to contribute this training scripts into `examples/elastic/tensorflow2` folder, since we don't have a keras+tf2 example today, as users might run into similar issues as you have.

If you agree, I will make a PR, put you as co-author for it. Please let us know.

Please take a check of the change and let me know if you still have issues. 


[train_gloo_fixed.py.zip](https://github.com/horovod/horovod/files/6013047/train_gloo_fixed.py.zip)

**Comment by czmrand** [comment]
@chongxiaocThanks for the update!!Okay with me to add as an example.Thanks,Chaim    On Friday, February 19, 2021, 11:47:45 PM GMT+2, chongxiaoc <notifications@github.com> wrote:  
 
 


@czmrand
I finally fixed the issue.
Please see the attached training scripts, and the diff is as below:
(hvd-py3.7) chongxiaoc@chongxiaoc-C02DL7WPMD6R hvd_sandbox % diff train_gloo.py train_gloo_fixed.py
62c62,66
<     state = hvd.elastic.KerasState(model, batch=100, epoch=0)
---
>
>     # Horovod: initialize model and optimizer state so we can synchronize across workers
>     model.fit(ds, steps_per_epoch=1, epochs=1, callbacks=None)
>
>     state = hvd.elastic.KerasState(model, batch=0, epoch=0)
66,67d69
<         hvd.elastic.CommitStateCallback(state),
<         hvd.elastic.UpdateBatchStateCallback(state),
68a71,72
>         hvd.elastic.UpdateBatchStateCallback(state),
>         hvd.elastic.CommitStateCallback(state),
71d74
<
74c77
<         model.fit(ds, steps_per_epoch=100, epochs=100, callbacks=callbacks)
---
>         state.model.fit(ds, steps_per_epoch=100, epochs=100-state.epoch, callbacks=callbacks)

Some import points you were missing:
   
   - Keras/TF2 has a lazy initialization for optimizer weights (https://sourcegraph.com/github.com/tensorflow/tensorflow@v2.4.1/-/blob/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L351:10), so hovorod has to initialize optimizer state by running one step of training so it can synchronize across all ranks, you can refer it by checking tensorflow2_mnist_elastic.py https://github.com/horovod/horovod/blob/f3af98649a26f4f3725fcab6a9bd3e8d29d7ffd2/examples/elastic/tensorflow2/tensorflow2_mnist_elastic.py#L70

Your above example hangs there, because the new joined rank has an empty optimizer state to sync with existing rank 0 and rank1, i.e.,
[0]:2: [HorovodBroadcast_Adam_conv1_bn_beta_m_0, HorovodBroadcast_Adam_conv1_bn_beta_v_0, HorovodBroadcast_Adam_conv1_bn_gamma_m_0, HorovodBroadcast_Adam_conv1_bn_gamma_v_0, HorovodBroadcast_Adam_conv1_conv_bias_m_0, HorovodBroadcast_Adam_conv1_conv_bias_v_0 ...]

There are all optimizer variables.
   
   - Also I adjusted a few places of using elastic state and callback functions order, very minor changes.

Now the training script is running fine on my system by injecting a failure of using removal hosts, and also adding hosts back into training works fine, no hanging anymore.

I would like to take this chance of fixing your example, to contribute this training scripts into examples/elastic/tensorflow2 folder, since we don't have a keras+tf2 example today, as users might run into similar issues as you have.

If you agree, I will make a PR, put you as co-author for it. Please let us know.

Please take a check of the change and let me know if you still have issues.

train_gloo_fixed.py.zip

**Comment by chongxiaoc** [comment]
Closed as tf2/keras elastic example is commited.
